src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,305 @@
+            private async ValueTask StartWritingByteValueInChunksAsync(ReadOnlyMemory<byte> value)
+            {
+                // Process the input value in chunks
+                for (int i = 0; i < value.Length; i += chunkSize)
+                {
+                    int remainingBytes = Math.Min(chunkSize, value.Length - i);
+                    bool isFinalBlock = false ;
+
+                    // Take a chunk of bytes from the input value.
+                    ReadOnlyMemory<byte> chunk = value.Slice(i, remainingBytes);
+
+                    // Get unprocessed bytes from the buffer.
+                    ReadOnlyMemory<byte> bytesNotProcessedFromPreviousChunk = buffer;
+
+                    // If the buffer is not empty, then we copy the bytes from the buffer
+                    // to the current chunk being processed.
+                    if (!bytesNotProcessedFromPreviousChunk.IsEmpty)
+                    {
+                        // Create a new combined array to hold bytes from both previous and current chunks.
+                        byte[] combinedArray = new byte[bytesNotProcessedFromPreviousChunk.Length + chunk.Length];
+
+                        // Copy bytes from bytesNotProcessedFromPreviousChunk to the combined array
+                        bytesNotProcessedFromPreviousChunk.Span.CopyTo(combinedArray);
+
+                        // Copy bytes from chunk to the combined array, starting from the end of bytesNotProcessedFromPreviousChunk
+                        chunk.Span.CopyTo(combinedArray.AsSpan(bytesNotProcessedFromPreviousChunk.Length));
+
+                        WriteChunk(combinedArray, isFinalBlock);
+                    }
+                    else
+                    {
+                        WriteChunk(chunk.Span, isFinalBlock);
+                    }
+
+                    // Flush the writer if the buffer threshold is reached
+                    await this.jsonWriter.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+                }
+            }
+
+            /// <summary>
+            /// Writes a chunk of data, encoding it using Base64, and updates the buffer with any unprocessed bytes.
+            /// </summary>
+            /// <param name="chunk">The chunk of data to be encoded and written.</param>
+            /// <param name="isFinalBlock">A boolean indicating whether this is the final block of data.</param>
+            private void WriteChunk(ReadOnlySpan<byte> chunk, bool isFinalBlock)
+            {
+                // Encode the current chunk using Base64 and write it
+                this.jsonWriter.Base64EncodeAndWriteChunk(chunk, isFinalBlock, out int bytesNotWrittenFromPreviousChunk);
+
+                if (bytesNotWrittenFromPreviousChunk > 0)
+                {
+                    // Update the buffer with unprocessed bytes from the current chunk.
+                    buffer = chunk.Slice(chunk.Length - bytesNotWrittenFromPreviousChunk).ToArray();

```

Comment:
`ToArray()` allocates a new array. Do we need to allocate a new array each time? I do see the convenience in this approach over allocating once and reusing the same array over and over: you don't have to figure out what size is good enough, you don't have to keep track of how much was put in the buffer or to clear the buffer before each write, etc.

But imagine that a lot of data is being streamed, we'll have a high density of small array allocations resulting from this that could potentially be avoided. I also imagine that since thus buffer is used to store leftover characters from the encoding, it might never need to be resized. I think we can accurately estimate a small upper bound for the buffer size.  I assume 8 bytes or even 4 could suffice, but subject to testing to ensure that's correct. Then you can use a separate variable to keep track of how much data has been written to the buffer

Reply:
Based on [some rudimentary testing](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRj1UwAJUCA6AIQgDM6BTMAZ0oBVGAPYAbmLIUO3PnjwAjAJ7BGAbQC6pOAENgy0gF5SURgHdSUmbIAs8/rjoB7MKQAUASyjBS9zaUy8XpADxLVyygAZRigAc2AAC097REQASjwAbzxSVL81WXtFLVtDRjiHUkRSAjjzAF8xXABlAAdlKG88gD4XKFqIZy0VNXM8hVIwRhYIABsu7T0DaTlqZRZGADZjSgBxRmAAWWUuAFEoAGNLOEY4dksAVWA6AA5gsMjbHoD78Ii4szw6hqaZ1stOh0JkMRuNzPsjicAIJQOAAdTA9hkDnanRgpABwCB6MxbWcRygIwAtqccZ08aRdIjgDJYC4WAAxRzKUbUUaWA4AaxApDoLIWZSqjmc0CpSNpbmeQRCb1IAFpSATiadzN9Gi1SLUQnBHKE3I4gZRqqN7AdGLYlRASXB0aLqbTBbgIcdGDD4dTzVrYbqyVjOkaTWbbGKaSE4r7FZZCVbSRjySHaej7Izmaz2VyecBIPlzFUAPR5/qKOiptkczkAJWGYwmOn0/TmC2Waw22y4ABFGJDTgywJYiVdbq9HpgPuYC87obCEUjzVCwGBlJJKLsibVgJJvNQ/rZw7zS+nK9XxhHLdaIwmQkmU1AWWWM6QsxAc3g8wAqL71dV/fe30YAeUBcktExQ1jVNc1MHRS8oEdCgyBLP9qizXU3EnXVKAudgGRuFtgGQxEwlsRC70PKtQWAdEoMpe0w3MCgAE5bAAIhI0ZyAITBmLgggmLYgjdUdN88yqYhjFISdXWnD1bCrZQ4H/KBRkkNVfhkVoDTRUhVI1UCtNxYVI2jc842cQyYPRcRLEsdjkyZP9705BJcGSXA0lIRslhWSTzkHG4UWxUzAtxM9Y1xCz6Xs0jyx5OyD3LMpSBSNJGJYu1xRCaYZBYUhEkwcpmKTVFgGlB4InlIzlTgR13J0n8YMAv0Jj0krwKDaiYJqlLOJojKoAEsI0MOY4MKwnC8IG0Jg1oqBGqBLrUlSmDJsdSpcCqIA), it seems we could have most 2 unwritten bytes. Like inputLength % 3 bytes will be unwritten unless you're on the last block. But this is subject to verification.